src/lib/x509/x509_ext.cpp

```diff
@@ -868,6 +877,831 @@ const std::string& TNAuthList::Entry::telephone_number() const {
+      ASIdentifierChoice asnum;
+      as_obj_ber.decode(asnum);
+      m_asnum = asnum;
+
+      const BER_Object rdi_obj = seq_dec.get_next_object();
+      const ASN1_Type rdi_type_tag = rdi_obj.type_tag();
+      if(static_cast<uint32_t>(rdi_type_tag) == 1) {
+         BER_Decoder rdi_obj_ber = BER_Decoder(rdi_obj);
+         ASIdentifierChoice rdi;
+         rdi_obj_ber.decode(rdi);
+         m_rdi = rdi;
+      } else if(rdi_type_tag != ASN1_Type::NoObject) {
+         throw Decoding_Error(fmt("Unexpected type for ASIdentifiers rdi: {}", static_cast<uint32_t>(rdi_type_tag)));
+      }
+   }
+
+   // just an rdi
+   if(elem_type_tag == 1) {
+      BER_Decoder rdi_obj_ber = BER_Decoder(elem_obj);
+      ASIdentifierChoice rdi;
+      rdi_obj_ber.decode(rdi);
+      m_rdi = rdi;
+      const BER_Object end = seq_dec.get_next_object();
+      const ASN1_Type end_type_tag = end.type_tag();
+      if(end_type_tag != ASN1_Type::NoObject) {
+         throw Decoding_Error(
+            fmt("Unexpected element with type {} in ASIdentifiers", static_cast<uint32_t>(end_type_tag)));
+      }
+   }
+
+   seq_dec.end_cons();
+}
+
+void ASBlocks::ASIdentifierChoice::encode_into(Botan::DER_Encoder& into) const {
+   if(m_as_ranges.has_value()) {
+      into.start_sequence().encode_list(m_as_ranges.value()).end_cons();
+   } else {
+      into.encode_null();
+   }
+}
+
+ASBlocks::ASIdentifierChoice::ASIdentifierChoice(const std::optional<std::vector<ASIdOrRange>>& ranges) {
+   m_as_ranges = sort_and_merge_ranges<ASIdOrRange>(ranges);
+}
+
+void ASBlocks::ASIdentifierChoice::decode_from(Botan::BER_Decoder& from) {
+   const ASN1_Type next_tag = from.peek_next_object().type_tag();
+
+   if(next_tag == ASN1_Type::Null) {
+      m_as_ranges = std::nullopt;
+   } else if(next_tag == ASN1_Type::Sequence) {
+      std::vector<ASIdOrRange> as_ranges;
+      from.decode_list(as_ranges);
+
+      m_as_ranges = sort_and_merge_ranges<ASIdOrRange>(as_ranges);
+   } else {
+      throw Decoding_Error(fmt("Unexpected type for ASIdentifierChoice {}", static_cast<uint32_t>(next_tag)));
+   }
+}
+
+void ASBlocks::ASIdOrRange::encode_into(Botan::DER_Encoder& into) const {
+   if(m_min == m_max) {
+      into.encode(static_cast<size_t>(m_min));
+   } else {
+      if(m_min >= m_max) {
+         throw Encoding_Error("AS range numbers must be sorted");
+      }
+      into.start_sequence().encode(static_cast<size_t>(m_min)).encode(static_cast<size_t>(m_max)).end_cons();
+   }
+}
+
+void ASBlocks::ASIdOrRange::decode_from(BER_Decoder& from) {
+   const ASN1_Type next_tag = from.peek_next_object().type_tag();
+
+   size_t min;
+   size_t max;
+
+   if(next_tag == ASN1_Type::Integer) {
+      from.decode(min);
+      m_min = static_cast<uint32_t>(min);
+      m_max = m_min;
+   } else if(next_tag == ASN1_Type::Sequence) {
+      from.start_sequence().decode(min).decode(max).end_cons();
+      m_min = static_cast<uint32_t>(min);
+      m_max = static_cast<uint32_t>(max);
```

Comment:
Probably these should check for overflow. There are utils for this in `int_utils.h` (`checked_cast_to_or_throw`)